### ISSUE TITLE
Efficient way to populate an Uint8ClampArray from C++

### ISSUE THREAD

**Issue by neatudarius** [emscripten-core/emscripten#20114]
Hello,

Context: I am trying to allocate an Uint8ClampArray and pass it to some C++ library to directly populate it with data from GPU:
```
static auto emscriptenUint8ClampedArrayBuilder =
                                emscripten::val::global("Uint8ClampedArray");
 auto jsPixels = emscriptenUint8ClampedArrayBuilder.new_(bigSize);

skia_api(jsPixels.somehowGetInternalBuffer(), bigSize, ...);
```

I have few questions:
1. Is `jsPixels` allocated outside of the emscripten Heap? Or in the bug JS heap? My current emscripten heap is limited to 2GB. I really need to say from C++ "hey allocate a JS object outside of the emscripten heap because I will transfer it on other worker anyways".

2. I don't find a way to extract internal buffer from `Uint8ClampedArray` and convert it into a `uint8_t*`. Any hints?

3. I have this old conversion function:
```

auto convert(const uint8_t* pixels, int pixelSize) {
    //      Ideally the pixels should be written directly into a JS array to avoid the copy but
    //      there is no efficient way atm.
    auto jsPixels = emscripten::val::global("Uint8ClampedArray").new_(pixelSize);
    jsPixels.call<void>("set", emscripten::typed_memory_view(pixelSize), pixels));
    return jsPixels;
}
```
Here I have multiple buffers:
1. C++: use new to allocate the buffer
2. Pass buffer to Skia to copy data from GPU
3. Copy data from C++ to TS into Uint8ClampedArray.
Basically with Q1 + Q2 I'm trying to replace this logic and have a single allocation in JS and populate the buffer with a single copy in C++.

Best regards,
Darius

**Comment by brendandahl** [comment]
1) jsPixels is allocated on the JS heap
2) You can't expose the the underlying buffer from Uint8ClampedArray since it lives in JS. You'll have to allocate space for it in C++ and copy it in. This could also be done with a typed_memory_view.

It seems unlikely the skia api would accept an external JS typed array, so you'll have to do some copying somewhere in the pipeline.

If you want to do the minimal amount of allocating, I'd allocate the array in C++ and then expose it for reading/writing using a typed_memory_view.

**Comment by neatudarius** [comment]
thanks very much for clarification and hints!
